File: src/summary.js
import { getSessionMeta, setSessionMeta } from './config.js';
import { getSessionMessages } from './scanner.js';

/**
 * Generate a 1-line summary for a session using Anthropic API (Haiku).
 * Caches the result in config so it's only generated once.
 */
export async function generateSummary(session) {
  // Return cached summary if available
  const meta = getSessionMeta(session.sessionId);
  if (meta?.autoSummary) return meta.autoSummary;

  const apiKey = process.env.ANTHROPIC_API_KEY;
  if (!apiKey) return null;

  const messages = await getSessionMessages(session.filePath, 5);
  if (messages.length === 0) return null;

  try {
    const { default: Anthropic } = await import('@anthropic-ai/sdk');
    const client = new Anthropic({ apiKey });

    const response = await client.messages.create({
      model: 'claude-haiku-4-5-20251001',
      max_tokens: 100,
      messages: [
        {
          role: 'user',
          content: `Summarize what this coding session is about in exactly 1 short sentence (under 80 chars). Based on these user messages:\n\n${messages.map((m, i) => `${i + 1}. ${m}`).join('\n')}`,
        },
      ],
    });

    const summary = response.content[0]?.text?.trim();
    if (summary) {
      setSessionMeta(session.sessionId, { autoSummary: summary });
      return summary;
    }
  } catch {
    // API call failed, return null
  }

  return null;
}

/**
 * Generate summaries for all sessions that don't have one yet.
 */
export async function generateMissingSummaries(sessions, { onProgress } = {}) {
  const apiKey = process.env.ANTHROPIC_API_KEY;
  if (!apiKey) return;

  let done = 0;
  for (const session of sessions) {
    const meta = getSessionMeta(session.sessionId);
    if (meta?.autoSummary || meta?.description) {
      done++;
      continue;
    }

    const summary = await generateSummary(session);
    if (summary) session.autoSummary = summary;
    done++;
    onProgress?.(done, sessions.length);
  }
}

File: src-tauri/src/summary.rs
use anyhow::{anyhow, Result};
use serde_json::json;

const MODEL: &str = "claude-haiku-4-5-20251001";
const ENDPOINT: &str = "https://api.anthropic.com/v1/messages";

pub async fn generate_summary(api_key: &str, messages: &[String]) -> Result<String> {
    if messages.is_empty() {
        return Err(anyhow!("no messages to summarize"));
    }
    let joined = messages
        .iter()
        .enumerate()
        .map(|(i, m)| format!("[{}] {}", i + 1, m))
        .collect::<Vec<_>>()
        .join("\n");

    let prompt = format!(
        "Summarize what this coding session is about in exactly 1 short sentence (under 80 chars). Reply with just the sentence.\n\nUser messages:\n{}",
        joined
    );

    let client = reqwest::Client::new();
    let body = json!({
        "model": MODEL,
        "max_tokens": 120,
        "messages": [{"role": "user", "content": prompt}]
    });

    let resp = client
        .post(ENDPOINT)
        .header("x-api-key", api_key)
        .header("anthropic-version", "2023-06-01")
        .header("content-type", "application/json")
        .json(&body)
        .send()
        .await?;

    if !resp.status().is_success() {
        let status = resp.status();
        let text = resp.text().await.unwrap_or_default();
        return Err(anyhow!("anthropic api error {}: {}", status, text));
    }

    let val: serde_json::Value = resp.json().await?;
    let text = val
        .pointer("/content/0/text")
        .and_then(|v| v.as_str())
        .ok_or_else(|| anyhow!("missing content in response"))?;
    Ok(text.trim().to_string())
}
